clients/drcachesim/tests/burst_reattach.cpp

```diff
@@ -0,0 +1,158 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* This application links in drmemtrace_static and acquires a trace during
+ * a "burst" of execution and memory allocations (malloc() and free())
+ * in the middle of the application. It then detaches. Later it re-attaches
+ * and detaches again for several times.
+ */
+
+/* Like burst_static we deliberately do not include configure.h here */
+
+#include "dr_api.h"
+#include "drmemtrace/drmemtrace.h"
+#include <assert.h>
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <math.h>
+#include <stdlib.h>
+
+bool
+my_setenv(const char *var, const char *value)
+{
+#ifdef UNIX
+    return setenv(var, value, 1 /*override*/) == 0;
+#else
+    return SetEnvironmentVariable(var, value) == TRUE;
+#endif
+}
+
+// Test recording large values that require two entries.
+ptr_uint_t
+return_big_value(int arg)
+{
+    return (((ptr_uint_t)1 << (8 * sizeof(ptr_uint_t) - 1)) - 1) | arg;
+}
+
+static int
+do_some_work(int arg)
+{
+    static const int iters = 100;
+    double *val = new double; // libc malloc is called inside new
+    *val = arg;
+    for (int i = 0; i < iters; ++i) {
+        *val += *val + (double)return_big_value(i);
+    }
+    double temp = *val;
+    delete val; // libc free is called inside delete
+    return (temp > 0);
+}
+
+static void
+exit_cb(void *)
+{
+    const char *funclist_path;
+    drmemtrace_status_t res = drmemtrace_get_funclist_path(&funclist_path);
+    assert(res == DRMEMTRACE_SUCCESS);
+    std::ifstream stream(funclist_path);
+    assert(stream.good());
+    std::string line;
+    bool found_malloc = false;
+    bool found_return_big_value = false;
+    while (std::getline(stream, line)) {
+        assert(line.find('!') != std::string::npos);
+        if (line.find("!return_big_value") != std::string::npos)
+            found_return_big_value = true;
+        if (line.find("!malloc") != std::string::npos)
+            found_malloc = true;
+    }
+    assert(found_malloc);
+    assert(found_return_big_value);
+}
+
+int
+main(int argc, const char *argv[])
+{
+    if (!my_setenv("DYNAMORIO_OPTIONS",
+                   "-stderr_mask 0xc"
+                   " -client_lib '#;;-offline -record_heap"
+                   // Test the low-overhead-wrapping option.
+                   " -record_replace_retaddr"
+                   // Test large values that require two entries.
+                   " -record_function \"malloc|1&return_big_value|1\"'"))
+        std::cerr << "failed to set env var!\n";
+
+    for (int i = 0; i < 100; i++) {
```

Comment:
How much of the large runtime is from the do_some_work() size and we can thus easily shrink?

What if we set -max_global_trace_refs or other limits on data written?  Or is that not the bottleneck?

What if we ran -checklevel 1 or 0 -- not sure we want to do that though.

Reply:
Compared to the burst_malloc test, I minimized as much as I could the runtime of `do_some_work()` and stats reported (removed` -rstats_to_stderr`) and that resulted in the ~20% reduction that I initially reported. Beyond that I don't see more gains by reducing do_some_work. The bottleneck is the use of `record_function` or `record_heap`. Use of any those options increases the runtime by orders of magnitude. By removing these options and commenting out the assertions for finding malloc and return_big_value, the test only takes 15secs. So, setting up DR and attaching and detaching, and the work done have minimal overhead. Yet, without recording any functions the #4769 issue is not triggered.  Just keeping the recording for `malloc` (and still trigger #4769) but removing `-record_heap` and the recording of `return_big_value` results in 8% reduction in runtime which seems to be worth doing (included this change in the latest commit). Use of `-checklevel 1` additionally reduces runtime by 5% and setting it to 0 yields 12% reduction over not setting checklevel. So, the reduction with checklevel is probably not big enough to justify its use. Use of -max_global_trace_refs with various values (even very small ones) did not have much (if any) impact in performance.

Reply:
As discussed offline, -record_dynsym_only speeds up the test drastically: it seems it was symbol lookup combined with the option accumulation causing many many lookups that caused the slowness.